CMakeLists.txt

```diff
@@ -4,29 +4,131 @@
+project(uavcan C CXX)
+
+#=============================================================================
+# build options
+#
+# options are listed in a table format below
+set(opts
+    # name: type: default value: string options list : description
+    "CMAKE_BUILD_TYPE:STRING:RelWithDebInfo:Debug Release RelWithDebInfo MinSizeRel:Build type."
+    "CMAKE_CXX_FLAGS:STRING:::C++ flags."
+    "CMAKE_C_FLAGS:STRING:::C flags."
+    "UAVCAN_USE_CPP03:BOOL:OFF::Use cpp03 standard."
+    "UAVCAN_NO_ASSERTIONS:BOOL:OFF::Enable assertions."
+    "UAVCAN_USE_EXTERNAL_SNPRINTF:BOOL:OFF::Use external snprintf."
+    "UAVCAN_MEM_POOL_BLOCK_SIZE:STRING:48::memory pool block size"
+    "UAVCAN_STM32_TIMER_NUMBER:STRING:5::default timer"
+    "UAVCAN_MAX_NETWORK_SIZE_HINT:STRING:16::max network size"
+    "UAVCAN_PLATFORM:STRING:generic:generic linux stm32:Platform."
+    "UAVCAN_OS:STRING:generic:generic posix nuttx chibios:Operating system."
+    "UAVCAN_STM32_NUM_IFACES:STRING:2:1 2:Number of interfaces."
+    )
+foreach(_opt ${opts})
+    # arguments are : delimited
+    string(REPLACE ":" ";" _opt ${_opt})
+    list(GET _opt 0 _name)
+    list(GET _opt 1 _type)
+    list(GET _opt 2 _default)
+    list(GET _opt 3 _options)
+    list(GET _opt 4 _descr)
+    # options are space delimited
+    string(REPLACE " " ";" _options "${_options}")
+    # if a default has not already been defined, use default from table
+    if(NOT DEFINED DEFAULT_${_name})
+        set(DEFAULT_${_name} ${_default})
+    endif()
+    # option has not been set already or it is empty, set it with the default
+    if(NOT DEFINED ${_name} OR ${_name} STREQUAL "")
+        set(${_name} ${DEFAULT_${_name}})
+    endif()
+    # create a cache from the variable and force it to set
+    message(STATUS "${_name}\t: ${${_name}} : ${_descr}")
+    set("${_name}" "${${_name}}" CACHE "${_type}" "${_descr}" FORCE)
+    # if an options list is provided for the cache, set it
+    if("${_type}" STREQUAL "STRING" AND NOT "${_options}" STREQUAL "")
+        set_property(CACHE ${_name} PROPERTY STRINGS ${_options})
+    endif()
+endforeach()
+
+string(TOUPPER ${UAVCAN_OS} OS_UPPER)
+
+#=============================================================================
+# find libraries
+#
+if(${UAVCAN_OS} STREQUAL "nuttx")
+    find_path(NUTTX_EXPORT_DIR libs/libnuttx.a
+        PATHS
+        $ENV{HOME}/git/px4/Firmware/build_px4fmu-v2_default/px4fmu-v2/NuttX/nuttx-export
+        NO_DEFAULT_PATH
+        NO_CMAKE_FIND_ROOT_PATH
+        )
+    if(NOT NUTTX_EXPORT_DIR)
+        message(FATAL_ERROR "failed to find NUTTX_EXPORT_DIR, please set")
+    else()
+        message(STATUS "nuttx export: ${NUTTX_EXPORT_DIR}")
+    endif()
 endif()
 
-project(uavcan)
+#=============================================================================
+# set flags
+#
+
+add_definitions(
+    -DUAVCAN_MEM_POOL_BLOCK_SIZE=${UAVCAN_MEM_POOL_BLOCK_SIZE}
+    -DUAVCAN_MAX_NETWORK_SIZE_HINT=${UAVCAN_MAX_NETWORK_SIZE_HINT}
+    )
```

Comment:
Same question as above - why can't we pass these via preprocessor definitions? The application itself can configure these via `add_definitions()`, globally. There will be no name clashing because all preprocessor definitions used by libuavcan are prefixed with `UAVCAN_`.